[Security] Argon2i Password Encoder

zanbaldwin opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | WIP

Since the [libsodium RFC](https://wiki.php.net/rfc/libsodium) passed with flying colours, I'd like to kick start a discussion about adding Argon2i as a password encoder to the security component. The initial code proposal in this PR supports both the upcoming public API confirmed for PHP 7.2, and the [libsodium PECL extension](https://pecl.php.net/package/libsodium) for those below 7.2 (available for PHP 5.4+).

#### Concerns

- Should the test cover hash length? At the moment the result of Argon2i is 96 characters, but because the hashing parameters are included in the result (`$argon2i$v=19$m=32768,t=4,p=1$...`) this is not guaranteed.
- I've used one password encoder class because the result *should* be the same whether running natively in 7.2 or from the PECL extension, but should the logic be split out into separate private methods (like `Argon2iPasswordEncoder::encodePassword()`) or not (like in `Argon2iPasswordEncoder::isPasswordValid()`)? Since I can't really find anything concrete on Symfony choosing one way over another I'm assuming it's down to personal preference?

#### The Future

Whilst the libsodium RFC has been approved and the public API confirmed, there has been no confirmation of Argon2i becoming an official algorithm for `passhword_hash()`. If that is confirmed, then the implementation should *absolutely* use the native `password_*` functions since the `sodium_*` functions do not have an equivalent to the `password_needs_rehash()` function.

Any feedback would be greatly appreciated :smiley:

iltar commented:
With the constant changing of recommended algorithms, would it be an idea to provide a migration layer in Symfony? Currently people will have to build this themselves, but it could be a lot easier if Symfony would provide an extension point, especially now that we can rehash with native php functions.

zanderbaldwin commented:
That's a great idea since Symfony really should be promoting security best practices (though likely out of the scope of this PR so I'll have a go at starting that separately - if you've got any ideas on what direction that would take, what the config would look like, etc).

It's pretty much a given that Argon2i will be the next hashing algorithm supported by `password_hash()` (and therefore `password_needs_rehash()`) in PHP 7.2 but I haven't actually found anywhere that confirms it - either way it won't become `PASSWORD_DEFAULT` until at least 7.3 :cry:

stof commented:
> is a new password encoder considered a new feature (therefore should be against master) or a security improvement/patch (therefore should be against 2.8)?

to me, it is a new feature

stof commented:
> Is a MAX_PASSWORD_LENGTH of 72 appropriate? Argon2i is, in terms of execution duration, generally along the lines of BCrypt with a cost of 11.

the max password length of 72 for bcrypt is because the bcrypt algorithm does not support longer passwords at all (it truncates the input at 72 chars).
If Argon2i does not have a 72 chars restriction, we should not add it here (keeping the much longer restriction we have in all other encoders)

stof commented on src/Symfony/Bundle/SecurityBundle/Resources/config/security.xml:
```diff
@@ -14,6 +14,7 @@
         <parameter key="security.encoder.plain.class">Symfony\Component\Security\Core\Encoder\PlaintextPasswordEncoder</parameter>
         <parameter key="security.encoder.pbkdf2.class">Symfony\Component\Security\Core\Encoder\Pbkdf2PasswordEncoder</parameter>
         <parameter key="security.encoder.bcrypt.class">Symfony\Component\Security\Core\Encoder\BCryptPasswordEncoder</parameter>
+        <parameter key="security.encoder.argon2i.class">Symfony\Component\Security\Core\Encoder\Argon2iPasswordEncoder</parameter>
```
we don't use parameters for class names anymore

stof commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php:
```diff
@@ -85,6 +86,28 @@ public function testEncodePasswordPbkdf2()
+        if (PHP_VERSION_ID < 70200 && !extension_loaded('libsodium')) {
+            $this->markTestSkipped('Argon2i algorithm not available.');
+
+            return;
```
useless. ``markTestSkipped`` is throwing an exception to interrupt the execution of the test

stof commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,85 @@
+
+        if (PHP_VERSION_ID >= 70200) {
+            return $this->encodePasswordNative($raw);
+        } elseif (extension_loaded('libsodium')) {
```
just ``if``

stof commented:
Please rebase this on top of master and change the target branch of the PR

zanderbaldwin commented:
All done, thanks. I'll update the `MAX_PASSWORD_LENGTH` once I can find some concrete evidence that Argon2i does not have the same length limitations as BCrypt.

stof commented:
@zanderbaldwin if the algo does not explicitly says it truncates the input, it means it does not do it. you would have to look at the algo description.

stof commented:
another solution to detect it experimentally: hash a very long password, and then verify it with a string changing a single char in the string. and keep doing it, by increasing the altered index little by little. Once you find a case where the altered string is accepted in verify, it means you passed over the accepted input length.
Of course, you don't need to increase 1 by 1 (unless you have days to spend on your machine for that, or a botnet to distribute the work). You can start with a high value and increase by big numbers, until you find an accepted mismatch, and then bisect the length being the max.

zanderbaldwin commented:
Confirmed, I tested up to 8192 characters and there are no false-positives from truncating. More than enough for a `MAX_PASSWORD_LENGTH` of 4096.

jedisct1 commented:
You can safely use passwords up to 2^32 bits long.

derrabus commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,84 @@
+            throw new BadCredentialsException('Invalid password.');
+        }
+
+        if (PHP_VERSION_ID >= 70200) {
```
I'd prefer checking for features over versions whenever possible. In this case:
```
if (function_exists('sodium_crypto_pwhash_str'))
```
What do you think?

iltar replied:
I agree, feature checking is more reliable when you think of php on .net or hhvm.

zanderbaldwin replied:
Agree. I'll add that in today.
I'm also trying to find a definitive answer on whether `PASSWORD_ARGON2I` will be implemented for `password_hash()` which would be a better solution than `sodium_crypto_passwd_str()`, so that feature test could change.

stof commented on src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php:
```diff
@@ -94,11 +95,16 @@ protected function execute(InputInterface $input, OutputInterface $output)
 
         $encoder = $this->getContainer()->get('security.encoder_factory')->getEncoder($userClass);
         $bcryptWithoutEmptySalt = !$emptySalt && $encoder instanceof BCryptPasswordEncoder;
+        $argon2iWithoutEmptySalt = !$emptySalt && $encoder instanceof Argon2iPasswordEncoder;
```
I suggest adding a a marker interface on encoders not needing the salt externally, to avoid having to know them one by one (I have the same need for detecting them in FOSUserBundle 2 to avoid generating useless salts).
what do you think @symfony/deciders ?

zanderbaldwin replied:
+1 even though I'm not a decider; I wanted to do something like that but thought it would be adding too much for the scope of this PR.

Is that something I could implement here or separate PR?

stof replied:
let's do it separately indeed. But the introduction of this encoder makes it worth adding such feature.

xabbuh replied:
Good idea!

zanderbaldwin commented:
Hi, all requested changes have been made and I've finally got round to fixing the conflicts with the latest version of `master` :+1:

There will be conflicts with #21620 that I'll have to fix (most likely in this PR if #21620 gets merged first since it's been marked for `3.3`).

xabbuh commented on src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,61 @@
+
+    public function setUp()
+    {
+        if (PHP_VERSION_ID < 70200 && !extension_loaded('libsodium')) {
```
Isn't it possible to run the tests if libsodium is missing, but the  `sodium_crypto_pwhash_str()`, `sodium_memzero()`, and `sodium_crypto_pwhash_str_verify()` functions do exist?

zanderbaldwin replied:
Yes, absolutely. I did change from version-checking to feature-checking in `Argon2iPasswordEncoder` but I unfortunately either forgot to update it here, or it snuck in whilst rebasing (either way I missed it and shall update when I get home from work).

The `isPasswordValid()` method checks for the function that it uses, but neither `encodePassword()` or `isPasswordValid()` check for the existence of `sodium_memzero()` - is this something that should be done in case one of the check functions has been defined in user-land code? It's a pretty far off edge-case but should at least be aknowledged...

zanderbaldwin replied:
One solution could be to check the result of `ReflectionFunctionAbstract::isUserDefined`, though would this be adding too much overhead?

nicolas-grekas commented:
`src/Symfony/Component/Security/Core/Encoder/EncoderFactory.php` should also be updated to account for argon2

zanderbaldwin commented:
I've updated `EncoderFactory`, but I admit that I didn't quite comprehend the `EncoderFactoryTest` to modify it appropriately, if this needs updating can someone point me in the right direction please?

sstok commented:
EncoderFactoryTest doesn't need to be updated, both bcrypt and pbkdf2 are not listed in this test-class also.

sstok commented on src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,61 @@
+/**
+ * @author Elnur Abdurrakhimov <elnur@elnur.pro>
+ */
+class Argon2iPasswordEncoderTest extends \PHPUnit_Framework_TestCase
```
This needs to use `PHPUnit\Framework\TestCase`.

sstok replied:
This needs to use `PHPUnit\Framework\TestCase`.

chalasr commented:
> Whilst the libsodium RFC has been approved and the public API confirmed, there has been no confirmation of Argon2i becoming an official algorithm for passhword_hash(). If that is confirmed, then the implementation should absolutely use the native password_* functions since the sodium_* functions do not have an equivalent to the password_needs_rehash() function.

Confirmed it is: PHP 7.2 introduces `PASSWORD_ARGON2I` as algo for `password_rehash()` (see https://wiki.php.net/rfc/argon2_password_hash).

@zanderbaldwin What's the status of this PR? Could you finish it? Or do you want someone to take it over? I would like to get this in.

zanderbaldwin commented:
Sure, I'll get on this in the next few days.

Does anyone know a good way to determine if the constant `PASSWORD_ARGON2I` is user or core defined? I'd rather feature test than test `PHP_VERSION_ID >= 70200`.

chalasr commented:
@zanderbaldwin Just tried out, using `defined('PASSWORD_ARGON2I')` should do the job.
See https://3v4l.org/N8RCV

nicolas-grekas commented:
@zanbaldwin would you have time to finish this PR for the end of the week?

zanbaldwin committed:
Argon2i Password Encoder

Add the Argon2i hashing algorithm provided by libsodium as a core encoder in the Security component, and enable it in the SecurityBundle.
Credit to @chalasr for help with unit tests.

zanbaldwin commented:
I've made most of the changes. Some points I want to clarify are:

- `PASSWORD_ARGON2I` could be user-defined in versions of PHP less than 7.2 so I also do version-testing. But I hate version checking instead of feature checking, so does anyone know of a better way of doing this?
- There are three different ways to test for the presence of Sodium, do we need them all?
  - `PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')` uses the built in `password_hash()` functions.
  - `function_exists('sodium_crypto_pwhash_str')` uses the Sodium functions that are native in 7.2.
  - `extension_loaded('libsodium')` uses the Sodium extension for PHP versions less than 7.2.

I'm thinking of getting rid of the `function_exists('sodium_crypto_pwhash_str')` check, but while testing on `php:7.2.0RC2-alpine` Docker image I've found that the constant `PASSWORD_ARGON2I` is not defined (but it works completely fine on 3v4l.org). Thoughts?

zanbaldwin commented:
**Also**, GitHub's *Resolve conflicts* button says there is a conflict in [UserPasswordEncoderCommand.php](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/Command/UserPasswordEncoderCommand.php), but when I try to rebase `master` onto this branch I get conflicts in [CHANGELOG.md](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/CHANGELOG.md) and [UPGRADE-4.0.md](https://github.com/symfony/symfony/blob/master/UPGRADE-4.0.md).

Have I really used Git so little in the last few months I've forgotten how to resolve conflicts properly? :weary:

zanbaldwin commented:
Ah, I did `git rebase master` since that's the target branch.
Would that be the reason Travis is spitting out 48 failed tests?

nicolas-grekas commented:
3.4 is the branch where we add features, so the target branch should also be updated.

sstok commented:
Make a back-up first (`git branch 2.8-argon2i-encoder-bak`)
`git rebase --onto 3.4 master` and then push with force, don't forget to change the pull-request target branch.

stof commented:
should also be rebased once https://github.com/symfony/symfony/pull/21620 is merged to benefit from it

stof commented:
@zanbaldwin please rebase this as the SelfSaltingEncoderInterface is now merged.

chalasr commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,96 @@
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        $valid = false;
+        if (PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')) {
```
`\PHP_VERSION_ID` (same for other php version checks)

chalasr reviewed (commented):
Can you add a note in SecurityBundle and Security CHANGELOG files?

fabpot commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,96 @@
+
+        throw new \LogicException(
+            'Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.'
+        );
```
Should be on one line

fabpot commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php:
```diff
@@ -129,6 +153,23 @@ public function testEncodePasswordBcryptOutput()
+        if (!function_exists('sodium_crypto_pwhash_str')
+            && !extension_loaded('libsodium')
+            && !defined('PASSWORD_ARGON2I')
+        ) {
```
condition should be on one line.

fabpot commented on src/Symfony/Bundle/SecurityBundle/Tests/Functional/UserPasswordEncoderCommandTest.php:
```diff
@@ -69,6 +70,29 @@ public function testEncodePasswordBcrypt()
+        if (!function_exists('sodium_crypto_pwhash_str')
+            && !extension_loaded('libsodium')
+            && !defined('PASSWORD_ARGON2I')
+        ) {
```
condition should be on one line.

fabpot commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,96 @@
+    {
+        $hash = password_hash($raw, PASSWORD_ARGON2I);
+
+        return $hash;
```
Could return directly without an intermediate var.

fabpot commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,96 @@
+        } elseif (extension_loaded('libsodium')) {
+            $valid = !$this->isPasswordTooLong($raw) && \Sodium\crypto_pwhash_str_verify($encoded, $raw);
+            \Sodium\memzero($raw);
+        }
```
If there are no way to check if password is valid, we should throw an exception instead of just returning `false`. At least, that's what other encoder do as this makes debugging much easier.

fabpot commented on src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoderTest.php:
```diff
@@ -0,0 +1,65 @@
+    {
+        if (!function_exists('sodium_crypto_pwhash_str')
+            && !extension_loaded('libsodium')
+            && !defined('PASSWORD_ARGON2I')
```
should be on one line

xabbuh commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,93 @@
+            return $this->encodePasswordSodiumExtension($raw);
+        }
+
+        throw new \LogicException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
```
I wonder if we cannot already (also) do these checks inside `SecurityExtension` when the config is parsed? This would exclude WTF moments at runtime.

chalasr replied:
In the extension, it should be an `InvalidConfigurationException`

zanbaldwin replied:
Okay, I've made the changes in the Security component, I'll get onto putting extra checks in the SecurityBundle config parser :slightly_smiling_face:

fabpot replied:
same here, "install the libsodium extension" sounds better.

xabbuh commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,93 @@
+        throw new \LogicException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
+    }
+
+    private function encodePasswordNative($raw)
```
`private` methods should come after `public` ones

xabbuh commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,93 @@
+    {
+        if (\PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')) {
+            $valid = !$this->isPasswordTooLong($raw) && password_verify($raw, $encoded);
+        } elseif (function_exists('sodium_crypto_pwhash_str_verify')) {
```
I would use early returns instead of `elseif`

zanbaldwin commented:
I've made some changes, but committed them in [a separate commit](4b0d83d30a2531982e7508580b201c0c7b3c3ee5) (which can be squashed later).

- Update all exception messages regarding the Argon2i algorithm not being supported to be the same string.
- Add a [static helper method on the `Argon2iPasswordEncoder` class](https://github.com/zanbaldwin/symfony/blob/4b0d83d30a2531982e7508580b201c0c7b3c3ee5/src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php#L23-L28), all checks throughout the Security component and SecurityBundle use the same method. If this changes, it only needs to be updated in one place.
- Added a check in the [`SecurityExtension`](https://github.com/zanbaldwin/symfony/blob/4b0d83d30a2531982e7508580b201c0c7b3c3ee5/src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php#L613-L615). Is this the correct place to put it, or should the check be directly in the [`MainConfiguration`](https://github.com/zanbaldwin/symfony/blob/4b0d83d30a2531982e7508580b201c0c7b3c3ee5/src/Symfony/Bundle/SecurityBundle/DependencyInjection/MainConfiguration.php#L442) class using config validation callbacks?

Feedback welcomed, as always :+1:

xabbuh commented:
tests are currently failing (see the CI builds) :)

fabpot commented on src/Symfony/Bundle/SecurityBundle/CHANGELOG.md:
```diff
@@ -14,6 +14,7 @@ CHANGELOG
  * deprecated HTTP digest authentication
  * deprecated command `acl:set` along with `SetAclCommand` class
  * deprecated command `init:acl` along with `InitAclCommand` class
+ * Added support for the new Argon2i password encoder.
```
minor: dot at the end should be removed

fabpot commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php:
```diff
@@ -607,6 +608,18 @@ private function createEncoder($config, ContainerBuilder $container)
+        // Argon2i encoder
+        if ('argon2i' === $config['algorithm']) {
+            if (!Argon2iPasswordEncoder::isSupported()) {
+                throw new InvalidConfigurationException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
```
install the libsodium extension?

fabpot commented on src/Symfony/Component/Security/CHANGELOG.md:
```diff
@@ -13,6 +13,7 @@ CHANGELOG
    the user will always be logged out when the user has changed between
    requests.
  * deprecated HTTP digest authentication
+ * Added a new password encoder for the Argon2i hashing algorithm.
```
same, no dot

fabpot commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,104 @@
+use Symfony\Component\Security\Core\Exception\BadCredentialsException;
+
+/**
+ * Argon2iPasswordEncoder uses the Argon2i hashing algorithm provided by libsodium.
```
I would remove the "provided by libsodium" part, which is an implementation detail.

fabpot commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,104 @@
+            return $valid;
+        }
+
+        throw new \LogicException('Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.');
```
same here, "install the libsodium extension" sounds better.

fabpot reviewed (approved):
with minor text comments

zanbaldwin commented:
That's because Travis installs the [`symfony/security`](https://packagist.org/packages/symfony/security) package as version `dev-master 45020b8` rather than using the Security component code of this PR.

xabbuh commented:
@zanbaldwin In fact, the code is right (at least on the PHP 5.5 build), but the Argon2i algorithm is not available. Tests should reflect this possibility.

chalasr commented on src/Symfony/Component/Security/Core/Tests/Encoder/Argon2iPasswordEncoderTest.php:
```diff
@@ -0,0 +1,62 @@
+{
+    const PASSWORD = 'password';
+
+    public function setUp()
```
should be protected

chalasr commented on src/Symfony/Bundle/SecurityBundle/Tests/DependencyInjection/CompleteConfigurationTest.php:
```diff
@@ -448,6 +448,10 @@ public function testEncoders()
+            'JMS\FooBundle\Entity\User7' => array(
+                'class' => 'Symfony\Component\Security\Core\Encoder\Argon2iPasswordEncoder',
+                'arguments' => array(),
+            ),
```
This is what breaks tests, as it runs on PHP 5.6. I suggest writing a dedicated test in SecurityExtensionTest, skipped if argon2i not supported.

zanbaldwin replied:
I'm not quite sure how to implement that - will need more details or be pointed in the right direction if I'm to implement that by end of play tomorrow, sorry!

chalasr replied:
I did it for you in https://github.com/symfony/symfony/pull/21604/commits/51be4e50cacdc44f7dd65982ea281532acd581bc :)

zanbaldwin replied:
Awesome, thank you very much @chalasr! Looking at that commit, it would have taken me a long time to figure that out!

chalasr reviewed (commented):
almost good :)

ogizanagi commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,104 @@
+        if (\PHP_VERSION_ID >= 70200 && \defined('PASSWORD_ARGON2I')) {
+            return !$this->isPasswordTooLong($raw) && password_verify($raw, $encoded);
+        }
+        if (function_exists('sodium_crypto_pwhash_str_verify')) {
```
Should be fully qualified too, considering https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048#issuecomment-332784692 ?

stof replied:
indeed, it should

zanbaldwin replied:
Sure, will fully-qualify those two functions too :)

ogizanagi commented on src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php:
```diff
@@ -0,0 +1,104 @@
+
+            return $valid;
+        }
+        if (extension_loaded('libsodium')) {
```
Same

chalasr reviewed (approved):
All good

zanbaldwin commented:
Should I squash my five commits?

chalasr commented:
@zanbaldwin We can do it while merging, as you wish. 
@stof requested changes have been made, is this ok for you?

nicolas-grekas commented:
In fact yes, squashing is needed because we can't make it when several authors are in the history.

fabpot commented:
Thank you @zanbaldwin.
